runtime/autoload/man.vim

```diff
@@ -1,137 +1,252 @@
+  if !empty(s:tag_stack)
+    let tag = remove(s:tag_stack, -1)
+    execute tag['buf'].'b'
+    call cursor(tag['lnum'], tag['col'])
   endif
+endfunction
 
-  exec 'let s:man_tag_buf_'.s:man_tag_depth.' = '.bufnr('%')
-  exec 'let s:man_tag_lin_'.s:man_tag_depth.' = '.line('.')
-  exec 'let s:man_tag_col_'.s:man_tag_depth.' = '.col('.')
-  let s:man_tag_depth = s:man_tag_depth + 1
-
-  let editcmd = 'edit'
-  " Use an existing 'man' window, else open a new one.
-  if &filetype !=# 'man'
-    let thiswin = winnr()
-    wincmd b
-    if winnr() > 1
-      exec thiswin . 'wincmd w'
-      while 1
-        if &filetype ==# 'man'
-          break
-        endif
-        wincmd w
-        if thiswin == winnr()
-          break
-        endif
-      endwhile
+" find the closest man window above/left
+function! s:find_man() abort
+  if &filetype ==# 'man'
+    return 1
+  endif
+  let thiswin = winnr()
+  while 1
+    wincmd w
+    if &filetype ==# 'man'
+      return 1
+    elseif thiswin == winnr()
+      return 0
     endif
+  endwhile
+endfunction
 
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
-    endif
+let s:ref_char = '\%([^()[:space:]]\)'
+
+" returns the regex for a valid character in a manpage reference
+function! man#ref_char() abort
+  return s:ref_char
+endfunction
+
+let s:ref_chars = s:ref_char.'\+'
+
+" parses the page and sect out of 'page(sect)'
+function! s:parse_page_and_sect_fpage(fpage) abort
+  if a:fpage[0] ==# '-'
+    " to see why, disable this condition and try
+    " :echo systemlist('man 2>/dev/null -w -mandoc')
+    return ['', '']
+  elseif a:fpage =~# '^'.s:ref_chars.'$'
+    return [a:fpage, '']
+  elseif a:fpage !~# '^'.s:ref_chars.'('.s:ref_chars.')'
+    return ['', '']
   endif
+  let ret = split(a:fpage, '(')
+  let iret = split(ret[1], ')')
+  return [ret[0], tolower(iret[0])]
+endfunction
 
-  silent exec editcmd.' man://'.page.(empty(sect)?'':'('.sect.')')
+" parses the page and sect out of 'path/page.sect'
+function! s:parse_page_and_sect_path(path) abort
+  " regex for valid extensions that manpages can have
+  if a:path =~# '\.\%([glx]z\|bz2\|lzma\|Z\)$'
+    let tail = fnamemodify(fnamemodify(a:path, ':r'), ':t')
+  else
+    let tail = fnamemodify(a:path, ':t')
+  endif
+  let page = matchstr(tail, '^.\+\ze\.[^.]\+$')
+  let sect = matchstr(tail, '\.\zs[^.]\+$')
+  return [page, sect]
+endfunction
 
+function! s:read_page(sect, page)
+  let bname = 'man://'.a:page.(len(a:sect)?'('.a:sect.')':'')
+  if bufexists(bname)
+    execute 'buf '.bufnr(bname)
+    keepjumps 1
+    " needs to be called every time to set the window local options
+    " we cannot do for example `:set nonumber` because that completely unsets
+    " nonumber from the window. would much rather do `:setlocal nonumber` and
+    " then call it every time the buffer is loaded so that it is always set
+    " correctly for man.vim but whatever it was before for every other
+    " filetype; by default.
+    setlocal filetype=man
```

Comment:
In gitter we discussed setting window-local options _only_ if man.vim creates a new window or tab. It must not set window-local options if a new window/tab was not created.